src/operations/mission_controller/mission_controller/node_trackdrive_handler.py

```diff
@@ -1,59 +1,57 @@
+import time
+
 import rclpy
 from rclpy.node import Node
 from rclpy.publisher import Publisher
 
-import time
-
 from driverless_msgs.msg import Reset, Shutdown
 from geometry_msgs.msg import PoseWithCovarianceStamped
-from std_msgs.msg import Bool
+from std_msgs.msg import Bool, UInt8
 
 from driverless_common.shutdown_node import ShutdownNode
 
 
-class TrackdriveMission(ShutdownNode):
-    started: bool = False
+class TrackdriveHandler(ShutdownNode):
+    r2d: bool = False
     laps: int = 0
     last_lap_time: float = 0
 
     def __init__(self):
         super().__init__("trackdrive_logic_node")
-        self.create_subscription(Reset, "/system/reset", self.reset_callback, 10)
+        self.create_subscription(Bool, "/system/r2d", self.r2d_callback, 10)
         self.create_subscription(PoseWithCovarianceStamped, "/slam/car_pose", self.pose_callback, 10)
 
         self.shutdown_pub: Publisher = self.create_publisher(Shutdown, "/system/shutdown", 1)
-        self.lap_trig_pub: Publisher = self.create_publisher(Bool, "/system/lap_completed", 1)
-        
-        self.last_lap_time = time.time()
+        self.lap_trig_pub: Publisher = self.create_publisher(UInt8, "/system/laps_completed", 1)
 
-        self.get_logger().info("---Inspection mission node initialised---")
+        self.get_logger().info("---Trackdrive handler node initialised---")
 
-    def reset_callback(self, msg: Reset):
-        if not self.started:
-            self.started = True
-            self.lap_trig_pub.publish(Bool(data=False))
-            self.get_logger().info("Inspection mission started")
+    def r2d_callback(self, msg: Bool):
+        if not self.r2d and msg.data:
+            self.r2d = True
+            self.last_lap_time = time.time()
+            self.get_logger().info("Trackdrive mission started")
 
     def pose_callback(self, msg: PoseWithCovarianceStamped):
         # check if car has crossed the finish line (0,0)
         # get distance from 0,0 and increment laps when within a certain threshold
         # and distance is increasing away from 0,0
-        if self.started and msg.pose.pose.position.x < 0.3 and msg.pose.pose.position.y < 0.5:
+        if self.r2d and abs(msg.pose.pose.position.x) < 0.5 and abs(msg.pose.pose.position.y) < 2:
             if time.time() - self.last_lap_time > 20:  # 20 seconds at least between laps
```

Comment:
Funny how difficult it is to detect such a simple thing. 

I had an idea that might reduce the reliance on arbitrary values here. Store the previous pose's x value, then check the current pose's x value against it, it the previous was negative and the current is positive, and the y value is within the threshold shown here, then you have a new lap. This removes the need for x position checks against an arbitrary value and the minimum lap time check, if it works that is. Still leaves the y position check but the distance between blue and yellow cones is fairly regulated so I don't really consider it an arbitrary value.

Reply:
Yeah that's probably a good idea for X. The Y val was more troublesome if the car ran wide, so +-2m is actually wider than the track

Reply:
We could take the actual width of the track by subtracting one orange cone from the other, or we could store the y coordinates of two opposing orange cones and check that the car's y is between them.

Reply:
Do you want to try what I mentioned in my previous comment or are you happy with where it is here. Also I think you can remove the minimum lap time check here as when the previous x was negative and the current is positive, the last x is set to the positive x and the condition won't get hit again until next lap.

Reply:
Ah I completely missed your additional comment whoops.
Probably fine as it is, 2m is enough clearance so far. Also, the time is just a failsafe if theres any slight drift in the estimated pose which can slightly jiggle around before the car actually starts moving, sending it back and forth over the line.

Reply:
@supex0fan was that all g for approval btw?